`install.sh` script which makes the installation processes easier

qlrd opened the pull request:
### What is this PR for?

Fix https://github.com/vinteumorg/Floresta/issues/362

### What is the purpose of this pull request?

- [ ] Bug fix;
- [x] New feature;
- [ ] Docs update;
- [ ] Test;
- [ ] Other: <!-- Please describe it -->.

### Which aspect of floresta its being addresed?

- [ ] Blockchain;
- [ ] Nodes communication;
- [ ] User consumption;
- [ ] Utreexo accumulator; 
- [x] Other: installation script.

### Checklists

- [x] I've signed all my commits;
- [x] I ran `just lint`;
- [x] I ran `cargo test`;
- [x] I checked the integration tests;
- [x] I'm linking the issue being fixed by this PR (if any).

### Description

Added and `install.sh` script which makes the installation processes
easier (only for debian-like systems for now).

It installs apt-get dependencies, install rustup, build
florestad/floresta_cli and configure a proper systemd service in a
semi-customizable way.

It was also necessary to add some configurations to florestad.service so
that the service could run properly.

### Notes to the reviewers

It was inspired by @Davidson-Souza that already made scripts for Archlinux and with some DMs, a desire about make possible a procedure like `wget https://vinteum.org/floresta/install.sh | bash`.

jaoleal commented:
Nice, Concept ACK. 

Ill get for myself a debian distro to properly test this script but, looks very good.

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,321 @@
+# MAIN SECTION
+# Run all functions on sequence
+apt_install gcc build-essential pkg-config libssl-dev mold
+install_rustup
```
I think we would need a clean-up here. Maybe our user doesn't want rust installed in their systems?

Davidson-Souza reviewed (commented):
ConceptACK

Some preliminary comments.

Davidson-Souza commented:
Trying this on `Ubuntu 24 LTS`. If I already have `dialog` it works fine. If don't have dialog it shows `Checking if dialog is installed`, the terminal blinks and I get my prompt back (the script terminates and cleans my terminal). It does show an error, but for a fraction of a second.

If I already have dialog it doesn't happen

qlrd commented:
@lucad70 suggested, in person to add, on initial screen:

```
If you want to disclose a security vulnerability, please email Davidson Souza at me AT dlsouza DOT lol, using the PGP key [2C8E0F 836FD7D BBBB9E 9B2EF899 64EC3AB 22B2E3](https://blog.dlsouza.lol/assets/gpg.asc).
```

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1121 @@
+
+# func: validate_network
+#
+# this function validate the suposed network that florestad will run.
```
> This function validates the provided network. If an invalid network is provided, this function will show an error and abort the execution with exit 1

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1121 @@
+    echo "🔐 Generating SSL keys and certificates..."
+    ssl_dir="$florestaLib/ssl"
+    sudo mkdir -p "$ssl_dir"
+    sudo openssl req -x509 -nodes -days 365 -newkey rsa:2048 \
```
Is it possible to not have `openssl` installed on Debian-like systems/?

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1121 @@
+#
+# This function setup a proxy if user wants add one
+interactive_select_proxy() {
+  local proxy_regex="^((http|https|socks5):\/\/)?((([0-9]{1,3}\.){3}[0-9]{1,3})|\[?[a-fA-F0-9:]+\]?):[0-9]{1,5}$"  # Matches optional protocol, IPv4, IPv6, and port
```
We only expect a socks5 proxy, because this is for the p2p connections.

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1121 @@
+
+  while true; do
+    # Prompt user for proxy input
+    proxy=$(dialog --title "Floresta-Installer (Set Proxy)" \
```
> Provide a socks5 proxy address that floresta should use to open outgoing connections. This improves your privacy when running your node, but may make things slower. Examples of socks5 proxy are Tor and I2P.

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1121 @@
+  while true; do
+    # Prompt user for proxy input
+    connect=$(dialog --title "Floresta-Installer (Set a node to connect)" \
+                  --inputbox "We'll connect ONLY to this node. It should be an ipv4 address in the format <address>[:<port>]." \
```
On master, it can be an ipv4/6 and some dns address. Should we update this once we get a new release?

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1121 @@
+#
+# This function setup an address for the ZeroMQ server to listen to if user wants add one
+interactive_zeromq() {
+  local zeromq_regex="^((http|https):\/\/)?((([0-9]{1,3}\.){3}[0-9]{1,3})|\[?[a-fA-F0-9:]+\]?):[0-9]{1,5}$"
```
The zmq address is just the address. No prefix

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1121 @@
+  while true; do
+    # Prompt user for zeromq input
+    zeromq=$(dialog --title "Floresta-Installer (Set zeromq)" \
+                  --inputbox "Provide a zeromq address (Optional: http://domain:port, https://domain:port, IPv4:port or IPv6:port)" \
```
ZMQ is a lightweight and efficient message queue system used for Inter Process Communication. Floresta allows you to use this to get notified about new blocks. This option sets the socket that floresta will listen for ZMQ subscribers.

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1121 @@
+#
+# Ask if user want to use --cfilters or not
+interactive_cfilters() {
+  if dialog --title "Floresta-Installer" --yesno "Do you want to use 'cfilters'?\n\nThose filters let you query for chain data after IBD, like wallet rescan, finding a utxo, finding specific tx_ids. Will cause more disk usage" 15 60; then
```
Can we allow setting the start height for filters too?

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1121 @@
+#
+# Ask if user wants to use --assume-valid=<BLOCK_HASH> with validation
+interactive_ask_assume_valid() {
+  if dialog --title "Floresta-Installer" --yesno "Do you want to use 'assume-valid'?\n\nThis option assumes blocks before this height are valid. We still verify amounts, UTXOs and rewards, but skip script validation." 15 60; then
```
Assumes that **scripts** before this height are valid

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1121 @@
+# If user provided an invalid one, clean the wallet_xpubs list and back to interactive_ask_for_add_xpubs
+# to do all again.
+interactive_add_xpub() {
+  local xpub_regex="^xpub[A-Za-z0-9]{107}$"  # XPUBs are always 111 characters
```
xpubs may have other prefixes like `ypub` and `tpub`

Davidson-Souza reviewed (changes requested):
I've tested all options and with/without some requirements installed. It seems to work fine.

Leaving some comments, mostly related to the docs. 

Some of those options seems "Advanced". Is it hard to make an "Advanced Configuration" sub-menu? This should cause less confusion for newbies.

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1187 @@
+
+# func: interactive_main_menu
+#
+# Start asking if user wanst to install or uninstall.
```
Typo here in `wanst`

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1187 @@
+#
+# Ask if user wants to use --assume-valid=<BLOCK_HASH> with validation
+interactive_ask_assume_valid() {
+  if dialog --title "Floresta-Installer" --yesno "Do you want to use 'assume-valid'?\n\nThis option assumes that scripts before this height are validb." 15 60; then
```
`s/validb/valid/`

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1187 @@
+# Ask if user want to use SSL (Secure Sockets Layer) and if yes,
+# add --ssl-electrum-address, --ssl-cert-path and --ssl-key-path to florestad process.
+interactive_ssl() {
+  if dialog --title "Floresta-Installer" --yesno "Do you want to use SSL (Secure Sockets Layer)?" 8 45; then
```
> Do you want to use SSL (Secure Sockets Layer) for the Electrum Server?

> This is particularly important if you want to access the Electrum Server from a public untrusted network. This will add encryption and authentication to your service.

qlrd commented:
Added an advanced menu and simplified the main menu.

Should i squash to one commit or keep 2 commits?

Davidson-Souza commented:
If I go to some option like Wallet -> Add xpub and then select cancel, it shows me an error, but then just starts the installation process without letting me configure anything else

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1298 @@
+# func: download_floresta
+#
+# Download floresta from releases and check integrity
+# TODO: add signature verification when possible
```
Just as an FYI: Until 0.7.0 I did not sign the source code. This file is created by github and I didn't really care for it.

From 0.8.0 onward I will sign it as well

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1298 @@
+
+  sha256res=$(sha256sum $tarDest | awk -F" " '{ print $1 }')
+  sha256exp="733b1e2bcecfdf5ab552b81db428b125e6f8154cccfda69b5e746da7a4a0a322"
+  if [ ! "$sha256res" == "$sha256exp" ]; then
```
Something that could be done in a follow-up: even on error, we should call the clean-up methods. Even better would be to remove what was already copied over, so we don't leave garbage back from an unsuccessful install.

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1298 @@
+  fi
+
+  if [ -f "$florestaUserPath" ]; then
+    echo "🧹 Removing sysusers config: $florestaUserPath"
```
You have to call systemd-users after this, no?

Davidson-Souza reviewed (changes requested):
One extra round of review (hopefully one the last). Mostly about the setup, the interactive windows looks good.

qlrd commented:
> If I go to some option like Wallet -> Add xpub and then select cancel, it shows me an error, but then just starts the installation process without letting me configure anything else

fixed

Davidson-Souza commented:
On a fresh ubuntu install I've ran on the same thing using [91a0ede](https://github.com/vinteumorg/Floresta/commit/91a0ede21c025e3fab903312e468cbc382b7c25d). If I go into wallet > add xpub > cancel it shows invalid xpub, then invalid config and starts installing without asking me anything.

qlrd commented:
> On a fresh ubuntu install I've ran on the same thing using [91a0ede](https://github.com/vinteumorg/Floresta/commit/91a0ede21c025e3fab903312e468cbc382b7c25d). If I go into wallet > add xpub > cancel it shows invalid xpub, then invalid config and starts installing without asking me anything.

Reproduced and verified that the same behavior applies to the steps of adding descriptors and addresses.

Fixed.

I believe i should rebase it too, or no need?

Davidson-Souza commented:
> I believe i should rebase it too, or no need?

I don't think it's needed, it's up to you

When I select review and "yes", I'm being thrown back to the wallet configuration

qlrd commented:
> > > I believe i should rebase it too, or no need?
> > 
> > 
> > I don't think it's needed, it's up to you
> > When I select review and "yes", I'm being thrown back to the wallet configuration
> 
> Fixed (i believe), but now introduced another issue 😓. Can't go out of script when typing many ESCs.

fixed

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1367 @@
+interactive_add_xpub() {
+  local xpub_regex="^(xpub|ypub|zpub|tpub)[A-Za-z0-9]{107}$" # (X/Y/Z/T)PUBS are 111 chars
+
+  while true; do
```
If I select `Cancel`, it shows "Invalid Descriptor" and open the xpub input again

Davidson-Souza commented:
Tried briefly in a Ubuntu VM and seems to work fine :tada: :partying_face:. I'll do a final code review, but we should be all-set.

Just a minor something, could you make a doc mention for this?

qlrd commented:
> Tried briefly in a Ubuntu VM and seems to work fine 🎉 🥳. I'll do a final code review, but we should be all-set.
> 
> Just a minor something, could you make a doc mention for this?

Sure!

Davidson-Souza commented:
The images could live inside `doc/assets`. It's clearer to keep them out of the root folder for docs

lucad70 reviewed (commented):
This is really good. 

A few things:
- I had an issue running the `curl` command to download and run the script.
- You need to update the README.md and `doc/README.md` to reflect this PR.
- Maybe we could check for an already installed version of `florestad`? Say the user already has florestad but tries to run the script again just to try and use the configuration GUI. 
- I have also added a few NITs to improve readability.

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1365 @@
+florestaTmpPath="/usr/lib/tmpfiles.d/florestad.conf"
+florestaUserPath="/usr/lib/sysusers.d/florestad.conf"
+florestaService="/usr/lib/systemd/system/florestad.service"
+distribution=$(uname -m)
```
Also, this seems to not be used anywhere.

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1365 @@
+# This function validates the provided network. If an invalid network is provided, this function
+# will show an error and abort the execution with exit 1
+#
+# Valid networks are:
+#   - bitcoin
+#   - signet
+#   - testnet
+#   - regtest
```
testnet4?

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1365 @@
+# Install rustup with the recommended procedure (https://sh.rustup.rs)
+install_rustup() {
+    echo "🦀 Checking for rustup."
+    haveRust=$(which rustup)
```
If we have installed `cargo` through `apt`, I believe we won't have `rustup`. But we should use any `cargo` available in the system. So I think this should be:

```suggestion
    haveRust=$(which cargo)
```
instead

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1365 @@
+        export PATH=/home/$me/.cargo/bin:$PATH
+        curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | bash -s -- -y --no-modify-path
+
+        # If this script run with sudo, it will will make '.cargo'
+        # dir be owned by root. Change recursively this so the user will be the owner
```
We should probably just bail if we are called with `sudo`. It's too dangerous.

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1365 @@
+}
+
+# Rust cleanup
+# Remove toolchai and directories built within
```
nit: toochain

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1365 @@
+        echo "❌ Failed to modify /home/$me/.profile. Restoring /home/$me/.profile"
+        mv ~/.profile.bak ~/.profile
+        exit 1
+    fi
```
On success, maybe remove .profile.bak?

Davidson-Souza commented on contrib/install.sh:
```diff
@@ -0,0 +1,1365 @@
+    setup_service
+    download_floresta
+    build_floresta
+    cleanup_apt gcc build-essential pkg-config libssl-dev mold
```
One thing we can do as a follow-up is getting a bit smarter here. Because, what if the user already has `gcc` or `build-essential` installed? We will uninstall it anyway?
